강원대 FE 안성현 - 1단계 선물하기 메인 API 구현하기

anseonghyeon opened the pull request:
1. /api/themes API를 사용하여 테마 목록을 완성했습니다
2. 테마 목록의 데이터를 불러오는 동안 로딩 화면을 만들었습니다
3. 데이터가 없거나, 에러가 발생하면 선물 테마 섹션이 보여지지 않게했습니다
4. /api/products/ranking API를 사용하여 실시간 급상승 선물 랭킹 섹션을 만들었습니다
5. API 명세를 살펴보고 각 필터 선택 시 해당 필터에 맞는 API를 재 요청하도록 했습니다
6. 랭킹 섹션의 데이터를 불러오는 동안 보여줄 로딩 화면을 만들었습니다
7. 보여 줄 상품 목록이 없을경우 상품 목록이 없다는 문구를 만들었습니다

anseonghyeon committed:
feat(theme): theme api 연동과 로딩 스피너 추가

anseonghyeon committed:
feat(theme): api로 불러온 데이터가 없거나 에러가 나면 선물 테마 섹션이 보이지 않게함

anseonghyeon committed:
feat(ranking): 필터 선택시 조건에 맞는 API 요청하도록 코드를 수정

anseonghyeon committed:
feat(ranking): 랭킹 상품들 로딩중일때 스피너 화면 추가

anseonghyeon committed:
feat(ranking): 상품 목록이 없거나 에러일때 상품이 없습니다 뜨게함

anseonghyeon committed:
refactor(component): FriendSelector, GiftCategorySelector, PromoBanner, RealtimeGiftRank 컴포넌트들을 components/ 디렉토리에서 components/Home 디렉토리로 이동함

anseonghyeon committed:
Order페이지를 분리하려고 시도하다 실패해서 다시 restore 했습니다

anseonghyeon committed:
refactor(ranking): 실시간 랭킹 item을 클릭해서 Order페이지 진입시 로그인 여부 검증을 ProtectedRoute를 사용한 검증으로 리팩토링함

anseonghyeon committed:
refacator(GiftCategorySelector): 기존의 중첩 삼항연산자 구조를 if분을 사용한 조건랜더링으로 개선함

anseonghyeon committed:
refactor(RealtimeGiftRank): jsx에 사용된 중첩된 삼항 연산자를 if문으로 변경하여 구조 개선함

terajh commented on src/components/Home/GiftCategorySelector.tsx:
```diff
@@ -0,0 +1,137 @@
+    };
+
+    fetchThemes();
+  }, []);
```
[고민포인트] 지금 구현해주신 형태도 충분히 재사용 가능한 구조라고 생각합니다. 다만 GiftCategorySelector에서 API 요청 이후 에러 및 로딩 상태를 처리하는 일련의 과정이 RealtimeGiftRank에서도 동일하게 반복되고 있는 것 같아요.

이처럼 중복된 로직이 있다는 건, API 요청 후 반환된 promise를 기반으로 에러 여부에 따라 데이터를 처리하는 흐름 자체도 공통화할 수 있다는 의미일 수 있을 것 같습니다. 이런 흐름을 어떻게 잘 추상화하고 재사용할 수 있을지 한 번 고민해보시는 걸 추천드립니다 😊

anseonghyeon replied:
이 부분은 제가 추후 axios를 사용하는 부분이 더 많아지면 추상화 해서 커스텀훅으로 리팩토링해도 될까요? 지금 Order페이지 분리와 로그인시 userInfo를 관리하는 부분도 수정하고있어서
axios를 사용하는부분이 더 나오면 커스텀훅으로 리팩토링 해보겠습니다

terajh commented on src/components/Home/GiftCategorySelector.tsx:
```diff
@@ -0,0 +1,137 @@
+  useEffect(() => {
+    const fetchThemes = async () => {
+      try {
+        const response = await axios.get('http://localhost:3000/api/themes');
```
[의견] url을 const로 정의해 한곳에서 모아 관리하면,
추후 사용되는 API가 많아졌을 때 관리 포인트가 줄어들어 유지보수 측면에서 훨씬 효율적일 것 같습니다! 😊

terajh commented on src/components/Home/RealtimeGiftRank.tsx:
```diff
@@ -275,44 +363,7 @@ function RealtimeGiftRank() {
-        {isCollapsed ? '접기' : '더보기'}
-        </ExtraBtn>
-      </ExtraBtnWrapper>
+      <RealtimeRankItemWrapper selectedGroup={selectedGroup} selectedType={selectedType}/>
```
[의견] RealtimeRankItemWrapper 컴포넌트의 역할이 추가되었다보니 RealtimeGiftRank 파일이 꽤 방대해진것 같아요. 잘 분리해주신 김에 RealtimeRankItemWrapper 를 따로 컴포넌트로 분리해주어도 괜찮겠네요!

terajh reviewed (changes requested):
안녕하세요 성현님! 1단계 미션 고생 많으셨습니다 😊

PR description에 잘 정리해주신 것처럼, 전반적으로 요구사항을 충실히 반영해주셨고
커밋도 깔끔하게 단위별로 잘 나눠주셔서 리뷰하는 데 큰 도움이 되었습니다. 정말 수고 많으셨어요! 👏🏻👍🏻

이번 PR에는 간단한 수정 요청 코멘트 2건과 함께, [고민해보시면 좋을 포인트 1가지](https://github.com/next-step/react-gift-product-list/pull/77#discussion_r2210819061)를 함께 남겨보았습니다. 관련하여 수정 및 고민해보신 뒤 반영해주시면 좋을 것 같아, PR은 request changes로 드리겠습니다. 고민 포인트에 대해서는 직접 수정하셔도 좋고, 고민 과정을 정리해 주시면 확인하도록 하겠습니다.

혹시 과제 진행 중 궁금한 점이 있다면 PR description 이나 코멘트로 편하게 말씀주세요! 감사합니다 🙏🏻

anseonghyeon committed:
refacotr(env): API URL을 환경변수로 분리하여 유지보수가 용이하게함

anseonghyeon committed:
refacotr(RealtimeGiftRank)

: RealtimeRankItemWrapper컴포넌트를 별도의 파일로 분리함

anseonghyeon committed:
refactor(order): 주문 페이지의 컴포넌트들을 분리시킴

분리시키고 작동하긴하는데 Order페이지가 한덩어리일때 진짜 손도 댈수없어서 임시로 분리시켰습니다. 로직 개선과 사용되는 state와 props들의 선언 위치를 변경하고 modal컴포넌트도 한번더 분리해야할 것같습니다

terajh reviewed (approved):
성현님~ 1단계 미션 중 코멘트 드렸던 비대한 컴포넌티 분리에 대해 고민해보시고, 깔끔하게 리팩토링해주셨네요. 고생 많으셨습니다! 👍 수정해주신 코드들도 모두 잘 확인되었고, PR은 승인 후 머지하도록 하겠습니다. 

추가로 코멘트로 답변주신 axios 기반 api 모듈 리펙토링은 말씀주신것처럼 이후 단계진행시 고민해보시면 좋을 것 같습니다! 고생하셨습니다~~
